### ISSUE TITLE
reflectance and transmittance of diffracted orders of binary grating for oblique source don't sum to one

### ISSUE THREAD

**Issue by oskooi** [stevengj/meep#486]
As a validation of the mode-decomposition feature, which could also possibly be added to the test suite, the reflectance and transmittance of all diffracted orders for any non-lossy grating structure for an incident planewave source at any arbitrary angle must necessarily sum to one. The following script, which is an adaptation of the [binary grating tutorial example](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#diffraction-spectrum-of-a-binary-grating), sets up this test. For comparison, the script also computes the reflectance and transmittance directly via the Poynting flux. The total reflectance, transmittance, and their sum are displayed at the end of the simulation in two separate lines prefixed by `coeff:` (mode coefficients) and `flux:` (Poynting flux).

```py
mport meep as mp
import math
import cmath
import numpy as np

resolution = 40        # pixels/μm                                                                                                                                                              

dpml = 2.0             # PML thickness                                                                                                                                                          
dsub = 3.0             # substrate thickness                                                                                                                                                    
dpad = 3.0             # length of padding between grating and pml                                                                                                                              
gp = 10.0              # grating period                                                                                                                                                         
gh = 0.5               # grating height                                                                                                                                                         
gdc = 0.5              # grating duty cycle                                                                                                                                                     

sx = dpml+dsub+gh+dpad+dpml
sy = gp

cell_size = mp.Vector3(sx,sy,0)

# replace anisotropic PML with isotropic Absorber to attenuate parallel-directed fields of oblique source                                                                                       
abs_layers = [mp.Absorber(thickness=dpml,direction=mp.X)]

wvl = 0.5              # center wavelength                                                                                                                                                      
fcen = 1/wvl           # center frequency                                                                                                                                                       
df = 0.05*fcen         # frequency width                                                                                                                                                        

ng = 1.5
glass = mp.Medium(index=ng)

# rotation angle of incident planewave; CCW about Y axis, 0 degrees along +X axis                                                                                                                        
theta_in = math.radians(10.7)

# k (in vacuum) with correct length (plane of incidence: XY)                                                                                                                                    
k = mp.Vector3(math.cos(theta_in),math.sin(theta_in),0).scale(fcen)

symmetries = []
eig_parity = mp.ODD_Z
if theta_in == 0:
  k = mp.Vector3(0,0,0)
  symmetries = [mp.Mirror(mp.Y)]
  eig_parity += mp.EVEN_Y

def pw_amp(k,x0):
  def _pw_amp(x):
    return cmath.exp(1j*2*math.pi*ng*k.dot(x+x0))
  return _pw_amp

src_pt = mp.Vector3(-0.5*sx+dpml+0.3*dsub,0,0)
sources = [mp.Source(mp.GaussianSource(fcen,fwidth=df),
                     component=mp.Ez,
                     center=src_pt,
                     size=mp.Vector3(0,sy,0),
                     amp_func=pw_amp(k,src_pt))]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=abs_layers,
                    k_point=k,
                    default_material=glass,
                    sources=sources,
                    symmetries=symmetries)

refl_pt = mp.Vector3(-0.5*sx+dpml+0.5*dsub,0,0)
refl_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=refl_pt, size=mp.Vector3(0,sy,0)))

sim.run(until_after_sources=1000)

input_flux = mp.get_fluxes(refl_flux)
input_flux_data = sim.get_flux_data(refl_flux)

sim.reset_meep()

geometry = [mp.Block(material=glass, size=mp.Vector3(dpml+dsub,mp.inf,mp.inf), center=mp.Vector3(-0.5*sx+0.5*(dpml+dsub),0,0)),
            mp.Block(material=glass, size=mp.Vector3(gh,gdc*gp,mp.inf), center=mp.Vector3(-0.5*sx+dpml+dsub+0.5*gh,0,0))]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=abs_layers,
                    geometry=geometry,
                    k_point=k,
                    sources=sources,
                    symmetries=symmetries)

refl_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=refl_pt, size=mp.Vector3(0,sy,0)))
sim.load_minus_flux_data(refl_flux,input_flux_data)

tran_pt = mp.Vector3(0.5*sx-dpml-0.5*dpad,0,0)
tran_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=tran_pt, size=mp.Vector3(0,sy,0)))

sim.run(until_after_sources=mp.stop_when_fields_decayed(50, mp.Ez, src_pt, 1e-7))

r_flux = mp.get_fluxes(refl_flux)
t_flux = mp.get_fluxes(tran_flux)

r_nmode = np.floor((fcen*ng-k.y)*gp)-np.ceil((-fcen*ng-k.y)*gp) # number of reflected orders                                                                                                    
if theta_in == 0:
  r_nmode = r_nmode*0.5 # since eig_parity=mp.ODD_Z+mp.EVEN_Y which removes degeneracy                                                                                                          
Rsum = 0
for nm in range(int(r_nmode)):
  r_coeffs, r_vgrps, r_kpoints = sim.get_eigenmode_coefficients(refl_flux, [nm+1], eig_parity=eig_parity)
  Rmode = (abs(r_coeffs[0,0,1])**2)/input_flux[0]
  r_angle = math.acos(r_kpoints[0].x/(ng*fcen))
  ky = ng*fcen*math.sin(r_angle)
  if theta_in != 0:
    mnega = (ky+k.y)*gp           # mode number of negative diffracted order                                                                                                                    
    if mnega % int(mnega) < 1e-4: # determine if mode is a negative diffracted order                                                                                                            
      r_angle = -r_angle
  print("refl:, {}, {:.2f}, {:.8f}".format(nm,math.degrees(r_angle),Rmode))
  Rsum += Rmode

t_nmode = np.floor((fcen-k.y)*gp)-np.ceil((-fcen-k.y)*gp)       # number of transmitted orders                                                                                                  
if theta_in == 0:
  t_nmode = t_nmode*0.5 # since eig_parity=mp.ODD_Z+mp.EVEN_Y which removes degeneracy                                                                                                          
Tsum = 0
for nm in range(int(t_nmode)):
  t_coeffs, t_vgrps, t_kpoints = sim.get_eigenmode_coefficients(tran_flux, [nm+1], eig_parity=eig_parity)
  Tmode = (abs(t_coeffs[0,0,0])**2)/input_flux[0]
  t_angle = math.acos(t_kpoints[0].x/fcen)
  ky = fcen*math.sin(t_angle)
  if theta_in != 0:
    mnega = (ky+k.y)*gp           # mode number of negative diffracted order                                                                                                                    
    if mnega % int(mnega) < 1e-4: # determine if mode is a negative diffracted order                                                                                                            
      t_angle = -t_angle
  print("tran:, {}, {:.2f}, {:.8f}".format(nm,math.degrees(t_angle),Tmode))
  Tsum += Tmode

# display total reflectance, transmittance, and their sum of all diffracted orders computed using mode coefficients
print("coeff:, {:.6f}, {:.6f}, {:.6f}".format(Rsum,Tsum,Rsum+Tsum))

# display total reflectance, transmittance, and their sum computed directly from Poynting flux
Rflux = -r_flux[0]/input_flux[0]
Tflux =  t_flux[0]/input_flux[0]
print("flux:, {:.6f}, {:.6f}, {:.6f}".format(Rflux,Tflux,Rflux+Tflux))
```
Note that, following the discussion regarding oblique planewave sources in #471, the frequency width (`df`) is quite narrow (`0.05*fcen`) and also the `PML` has been replaced with an `Absorber`. Also, the sign of the angle of each diffracted order is determined following #476.

When the planewave is normally incident (`theta_in = 0`), the total reflectance and transmittance of all diffracted orders obtained via `get_eigenmode_coefficients` does indeed sum to one. Also, the mode-coefficient values agree with those from the corresponding Poynting flux:

```
coeff:, 0.045819, 0.926615, 0.972435
flux:, 0.048261, 0.950863, 0.999124
```
However, for an oblique source (`theta_in = 10.7`), the values are significantly different. Also, they do not  seem to be converging with increasing resolution:

**resolution=40**
```
coeff:, 0.029142, 0.623634, 0.652776
flux:, 0.056826, 0.942414, 0.999240
```

**resolution=50**
```
coeff:, 0.028733, 0.542246, 0.570979
flux:, 0.062082, 0.937214, 0.999296
```

**resolution=80**
```
coeff:, 0.022980, 0.842190, 0.865170
flux:, 0.059960, 0.940259, 1.000219
```

**resolution=100**
```
coeff:, 0.066360, 0.470629, 0.536989
flux:, 0.060384, 0.939963, 1.000347
```
Note the large change in the mode-coefficient values with resolution. Increasing both the `eig_tolerance` and `eig_resolution` parameters in the call to `get_eigenmode_coefficients` did not change the results. There seems to be another discretization effect at work; a potential bug in the implementation of `get_eigenmode_coefficients`?

**Comment by HomerReid** [comment]
The reflected and transmitted mode-expansion coefficients don't sum to 1 for general incident angles.

This can be seen already in the simplest case of a binary "grating" of zero height, i.e. `gh=0` in [this geometry](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#diffraction-spectrum-of-a-binary-grating), in which case we have just the simple problem of Fresnel scattering from a planar dielectric interface. Let the incident, reflected, and transmitted **E**-fields  respectively have amplitudes 1, *r*, *t*. 

There are two physical boundary conditions that must be satisfied at the interface:

(a) The tangential components of the **E** and **H** fields must be continuous. This yields two linear equations relating *r* and *t*.

(b) For a lossless dielectric the outgoing flux must equal the incoming flux. This yields a single quadratic relation relating *r* and *t*.

For the special case of normal incidence, the **E** and **H** fields are *purely* tangential, so the condition of tangential **E**-field continuity becomes simply *r+t=1*. More generally, for a nontrivial grating with multiple reflected and transmitted modes, this generalizes to a sum rule, &Sigma; *(r_n + t_n)*=1. This is the condition you have successfully replicated in the normal-incidence case. 

However, for oblique incidence, the **E** and **H** fields have both tangential *and* normal components, with only the former continuous across the interface. For example, in the Fresnel case the **E** fields look something like

+ incident:        *1 &times; (  cos(&theta;) **x** + sin(&theta;) **z** )*

+ reflected:       *r &times; ( -cos(&theta;) **x** + sin(&theta;) **z** )*

+ transmitted:   *t &times; (  cos(&theta;') **x** + sin(&theta;') **z** )* 

where &theta;' is the angle of transmission, related to the angle of incidence by Snell's law. So now the simple relation *r+t+1* that holds for normal incidence becomes the more complicated relation *(1-r)cos(&theta;) = t cos(&theta;')*---an equation that becomes more complicated still for a nontrivial grating.

On the other hand, the simple conservation-of-energy constraint on the *flux* must continue to hold for arbitrary incident angles. So the results you report don't strike me as obviously incorrect. 

Meanwhile, the failure of the mode-expansion coefficients to converge with resolution is a different story, perhaps related to the recent discussion of power flow normal to and parallel to PML layers in conjunction with issue #471.

**Comment by stevengj** [comment]
Try it with `gh = 0`.  MPB will still get diffracted orders, but their coefficients should all be approximately zero, except for the zero-th order coefficient which should be the given by the Fresnel formula.

**Comment by stevengj** [comment]
It looks like there might be a bug in the `eigenmode_amplitude` routine: the field needs to be multiplied by `exp(ik⋅x)`, since `maxwell_compute_e_from_d` only returns the Bloch envelope (the periodic part of the field).

(The code has been this way for a while — it was never a problem before because `exp(ik⋅x) = 1` for a source in a non-periodic medium or for normal incidence.  But we should fix this, not only for oblique sources, but to handle periodic media properly in general.)

In particular, it looks like the `eigenmode_amplitude` function should multiply its return value by `exp(2πi Gk⋅p)`, where `Gk` is `edata->Gk` and `p` is the `vec` argument of the function.

**Comment by oskooi** [comment]
The 1-pixel boundary bug fix in #499 (including the modification described in my [comment](https://github.com/stevengj/meep/pull/499#discussion_r220679099)) now produces correct results for the oblique-angle test described in this issue. 

Based on Steven's [comment](https://github.com/stevengj/meep/pull/499#issuecomment-424451976), the original script from above required changes to the definition of the `k_point` and the amplitude function (`amp_func`) of the planewave source:

```py
# k (in source medium) with correct length (plane of incidence: XY)                                                                                                           
k = mp.Vector3(math.cos(theta_in),math.sin(theta_in),0).scale(fcen*ng)
```
```py
def pw_amp(k,x0):
  def _pw_amp(x):
    return cmath.exp(1j*2*math.pi*k.dot(x+x0))
  return _pw_amp
```
With #499 and the above changes to the script, the sum of the reflectance and transmittance of all of the diffracted orders computed using `get_eigenmode_coefficients` sum to one. Also, these results are consistent wih the Poynting flux values. Finally, the results from `get_eigenmode_coefficients` converge with increasing resolution. 

The following are the results from the modified script at three resolutions: `25`, `50`, and `79`.

**resolution=25**
```
coeff:, 0.047251, 0.887981, 0.935232
flux:, 0.055556, 0.942197, 0.997753
```
**resolution=50**
```
coeff:, 0.056678, 0.921236, 0.977914
flux:, 0.060885, 0.938560, 0.999445
```
**resolution=79**
```
coeff:, 0.053758, 0.919943, 0.973701
flux:, 0.058235, 0.941545, 0.999780
```
We can therefore close this issue once #499 is merged.
